### ISSUE TITLE
Cannot Read Odrive estimated Theta/Gamma values over serial monitor

### ISSUE THREAD

**Issue by akrolic** [Nate711/StanfordDoggoProject#54]
We are using the Odrive v3.6 flashed with the custom firmware provided in this repository and are able to perform all of the open loop gaits with no issues. 
We are aware that this project was developed using the Odrive v3.5 but so far we have not had any issues with the custom firmware being incompatible with the latest Odrive hardware. However, we are running into a problem reading parameters from the Odrives and after a while troubleshooting the issue we could really use your help. Here is an overview of whats going on and what we know.

The goal is to obtain the theta and gamma values over the serial monitor.

We are printing the estimated and requested theta and gamma values using the debug.cpp file

```
void PrintODriveDebugInfo(struct ODrive odrv) {
    Serial.print(odrv.sp_theta, 2);
    Serial.print("\t");
    Serial.print(odrv.est_theta, 2);
    Serial.print("\t");
    Serial.print(odrv.sp_gamma, 2);
    Serial.print("\t");
    Serial.print(odrv.est_gamma, 2);
```

Which when printed on the serial monitor takes this format:
ODrive0:	 setpoint_theta, estimate_theta, setpoint_gamma, estimate_gamma

From the serial monitor after starting dance gait:
```
ODrive0:	0.00	0.00	0.00	0.00

DANCE
(f)req: 1.5000
step (l)ength: 0.0000
stance (h)eight: 0.1500
(d)own amplitude: 0.0500
(u)p amplitude: 0.0500
flight (p)roportion: 0.3500
(s)tep difference: 0.0000
Theta: 50.0000 0.5000
Gamma: 30.0000 0.5000

ODrive0:	0.00	0.00	1.41	0.00
```

The problem we are trying to solve is that we are not able to read back any of the estimated values from the Odrives.
We are have looked into the OdriveArduino.cpp file and  found the function responsible for constructing the messages from the serial data

```
int ODriveArduino::ParseDualPosition(char* msg, int len, float& th, float& ga) {
    // check if 1 byte for "P", 4 bytes holding encoder data, and 1 checksum byte were received
    if (len != 6) {
        return -1; // return -1 to indicate that the message length was wrong
    } else {
        // retrieve short from byte stream
        // remember that the first character is 'P'
        uint16_t th_16 = (msg[2] << 8) | msg[1];
        uint16_t ga_16 = (msg[4] << 8) | msg[3];
        uint8_t rcvdCheckSum = msg[5];

        // compute checksum
        uint8_t checkSum = 0;
        checkSum ^= msg[0]; // letter 'P'
        checkSum ^= msg[1];
        checkSum ^= msg[2];
        checkSum ^= msg[3];
        checkSum ^= msg[4];

        // DEBUG ONLY
        // Serial << "Comp checksum: " << (int)checkSum << " rcvd: " << (int)rcvdCheckSum << "\n";

        // if the computed and received check sums match then update the motor position variables
        if (checkSum == rcvdCheckSum) {
            // convert to float
            // TODO: figure out if casts are needed
            th = ((float) ((int16_t) th_16))/1000.0f;
            ga = ((float) ((int16_t) ga_16))/1000.0f;
            return 1;
        } else {
            // return -1 to indicate that the checksums didn't match
            return -1;
        }
    }
    return 1;
}
```

Then looking into uart.cpp the theta and gamma estimates would be read in when rx_state is in the READ_PAYLOAD case

```
case READ_PAYLOAD:
                //ADDED: Print statement
                Serial << "\nREAD_PAYLOAD\n";
                msg[msg_idx++] = c;
                if (msg_idx == payload_length) {
                    if (msg[0] == 'P') {
                        //ADDED: Print statement
                        Serial << "\nPROCESSING POSITION\n";
                        ProcessPositionMsg(msg, msg_idx, odrvSerial, odrvMsgOutput);
                    }
                    rx_state = IDLING;
                    msg_idx = 0;
                    payload_length = 0;
                    
                    #ifdef DEBUG_LOW
                        msg_end = micros();
                        Serial << "rcvd in: " << msg_end - msg_start << " in " << loop_iters << " loops\n";
                        // NOTE: As of code 10/14/18, the average receive time was 1100
                        // And number of loop executions to get a message was 0 or 1
                    #endif
                }
```

Finally the ProcessPositionMsg is defined in the same uart.cpp where it is calling the aforementioned ParseDualPosition function from the OdriveArduino.cpp

```
void ProcessPositionMsg(char* msg, int len, HardwareSerial& odrvSerial, struct MsgOutput& odrvMsgOutput) {
    #ifdef DEBUG_LOW
        for(int i=0; i<len; i++) {
            Serial << (int)msg[i] << "(" << msg[i] <<") ";
        }
        Serial << "\n";
    #endif

    #ifdef DEBUG_HIGH
        Serial <<  "rcv at: " << micros() << '\n';
    #endif

    float th,ga;
    int result = ODriveArduino::ParseDualPosition(msg, len, th, ga);

    // result: 1 means success, -1 means didn't get proper message
    if (result == 1) {
        // Update theta and gamma
        *(odrvMsgOutput.theta) = th;
        *(odrvMsgOutput.gamma) = ga;

        #ifdef DEBUG_LOW
            Serial << "Th,Ga: " << th << " " << ga << '\n';
        #endif

        // NOTE: it's possible that the feedback delay is wrong if an old
        // encoder reading took so long to come back that it came back after
        // another current command was sent; ie:
        // 1) Teensy sends command I1
        // 2) Teensy sends command I2
        // 3) Teensy receives reading E1
        // 4) Teensy thinks delay was E1-I2 when in reality it was E1 - I1
        // This problem won't happen if the delay is short and the control rate is small
        //
        latest_receive_timestamp = micros();
        global_debug_values.position_reply_time = latest_receive_timestamp - latest_send_timestamp;

        #ifdef DEBUG_HIGH
            Serial << "Done prs at: " << micros() << '\n';
            Serial << "Reply time (uS): " << global_debug_values.position_reply_time << "\n";
            // NOTE: As of 7/7/18 code, around 1500us from send to receive
        #endif
    } else {
        #ifdef DEBUG_LOW
            Serial.println("Parse failed. Wrong message length or bad checksum.");
        #endif
    }
}

```

We placed in some print statements in the ProcessSerial function to determine which state it is in and see when it reaches the READ_PAYLOAD case but it never did. We turned on the robot, sent the E command (performed the gait for about ~20 sec), sent the S command, and then when we turned off the ODrives the threads finally entered the while loop. The while statement condition is:
`while (odrvSerial.available()) `
So we are confused why the serial isn't available until the ODrives are turning off. 

As a side note, we also were not able to write any IMU values to the SD card. We have formatted the card to FAT32 and can see the created datalog##.csv files but they are 0kb (nothing is being written to the card and stored). I can create a separate issue for that though, just wanted to mention it here just incase.

**Comment by akrolic** [comment]
Hello again, we are still actively trying to fix this problem. It’s necessary for our project to be able to read back data from the odrives. We have tried several different changes to the code to try to find a work around but none have been successful. What we learned is that we are able to receive some rx bytes back only after the odrives have been turned off. The received serial messages are a series of special characters and noise, nothing like what should be read out.
Our teensy, xbee, and imu are all being powered from a separate battery than what the odrives are being powered by. We initially had issues when all devices were powered by one battery. A ground loop was completely preventing any open loop gait commands from being sent to the odrives over the serial connection from the teensy. 

I will make a separate post on the odrive forum to see if anyone else has experienced similar issues when controlling their motors. 

@CrazyAZ or @Nate711 do either of you recall if you were able to read the estimated theta and gamma messages back from the odrives? 

Thank you in advance for any support you can provide to our team.

**Comment by akrolic** [comment]
Issue has been resolved!

The problem was not on the teensy side, but rather on the Odrive firmware side. In hindsight this makes sense, but it was a bit of a journey to come to the realization. 

In /Odrive/Firmware/communication/ascii_protocol.cpp
The function 'ASCII_protocol_process_line' contains the following lines of code
```
    } else if (cmd[0] == 'C') { // dual current control
        float motor0_cur_sp, motor1_cur_sp;
        int result = parse_dual_current(cmd,len,motor0_cur_sp, motor1_cur_sp);

        if (result != 1) {
            respond(response_channel, use_checksum, "Failed on parse or checksum: ");
            respond(response_channel, use_checksum, cmd);
        } else {
            // set motor currents
            axes[0]->controller_.set_current_setpoint(motor0_cur_sp);
            axes[1]->controller_.set_current_setpoint(motor1_cur_sp);

            // IMPORTANT CODE: send motor encoder readings in response
            send_motor_positions(response_channel);
        }

    } else if (cmd[0] == 'P') { // coupled control
        float theta_sp, gamma_sp;
        int result = parse_dual_current(cmd,len,theta_sp, gamma_sp);

        const float MULTIPLIER = 1000.0;

        theta_sp /= MULTIPLIER;
        gamma_sp /= MULTIPLIER;

        if (result != 1) {
            respond(response_channel, use_checksum, "Failed on parse or checksum: ");
            respond(response_channel, use_checksum, cmd);
        } else {
            axes[0]->controller_.set_coupled_setpoints(theta_sp, gamma_sp);
            axes[1]->controller_.set_coupled_setpoints(theta_sp, gamma_sp);
        }
    } else if (cmd[0] == 'S') { // coupled control with gains
        float sp_theta, kp_theta, kd_theta;
        float sp_gamma, kp_gamma, kd_gamma;

        int result = parse_coupled_command(cmd, len, sp_theta, kp_theta, kd_theta, sp_gamma, kp_gamma, kd_gamma);

        if (result != 1) {
            respond(response_channel, use_checksum, "Failed to parse coupled command: ");
            respond(response_channel, use_checksum, cmd);
        } else {
            axes[0]->controller_.set_coupled_setpoints(sp_theta, sp_gamma);
            axes[0]->controller_.set_coupled_gains(kp_theta, kd_theta, kp_gamma, kd_gamma);
            
            axes[1]->controller_.set_coupled_setpoints(sp_theta, sp_gamma);
            axes[1]->controller_.set_coupled_gains(kp_theta, kd_theta, kp_gamma, kd_gamma);
        }
```

The version of the firmware we had was missing the "IMPORTANT CODE" which includes respond(). The definition for respond() is also contained in the same file and I will include it here for clarity.

```
void respond(StreamSink& output, bool include_checksum, const char * fmt, TArgs&& ... args) {
    char response[64];
    size_t len = snprintf(response, sizeof(response), fmt, std::forward<TArgs>(args)...);

    static uint8_t start_byte = 1;
    static uint8_t len_byte = 0;
    output.process_bytes((uint8_t*) &start_byte, 1, nullptr); // start byte
    output.process_bytes((uint8_t*) &len_byte,   1, nullptr); // byte indicating newline termination
    output.process_bytes((uint8_t*)response,   len, nullptr); // TODO: use process_all instead
    if (include_checksum) {
        uint8_t checksum = 0;
        for (size_t i = 0; i < len; ++i) {
            checksum ^= response[i];
        }
        len = snprintf(response, sizeof(response), "*%u", checksum);
        output.process_bytes((uint8_t*)response, len, nullptr);
    }
    output.process_bytes((const uint8_t*)"\n", 1, nullptr);
}
```

Now we are getting all the Theta and Gamma estimate values. Investigating this was a great learning experience for us, our next steps are to migrate all communications towards ROS
